pkg/apis/discovery/validation/validation_test.go

```diff
@@ -0,0 +1,329 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/apis/discovery"
+)
+
+func TestValidateEndpointSlice(t *testing.T) {
+	standardMeta := metav1.ObjectMeta{
+		Name:      "hello",
+		Namespace: "world",
+	}
+
+	nameHTTP := "http"
+	nameEmpty := ""
+	nameCaps := "ABCDEF"
+	nameChars := "almost_valid"
+	hostnameInvalid := "--invalid"
+	hostnameValid := "valid-123"
+	protoTCP := api.ProtocolTCP
+	addressTypeIP := discovery.AddressTypeIP
+	addressTypeOther := discovery.AddressType("Other")
+
+	testCases := map[string]struct {
+		isExpectedFailure bool
+		endpointSlice     *discovery.EndpointSlice
+	}{
+		"good-slice": {
```

Comment:
missing cases with multiple ports and multiple endpoints and multiple addresses.

All should probably have cases for 0, 1, and max, plus failure cases for max+1